[doc] Add documentation for Ray debugger (#11815)
ray-project/ray: doc/source/ray-debugging.rst

> Ray Debugging

Reviewer: ```suggestion
Ray Debugger
```

=============

Ray has a built in debugger that allows you to debug your distributed applications.

.. note::

>     It is currently an experimental feature and under active development. Interfaces are subject to change.

Reviewer: Can you please outline the key features? 

I'd like to know what this is/what features this has before I do a code-walkthrough.

Getting Started
---------------

> Take the following example that is modified from the Ray README:

Reviewer: ```suggestion
Take the following example:
```

.. code-block:: python

    import ray
    ray.init()

    @ray.remote
    def f(x):

>         ray.util.pdb.set_trace()

Reviewer: Is there a docstring for this?

Reviewer: If so, can you please provide it? 

This will be much more useful for users with an IDE.

Author: Thanks, I added a docstring!

        return x * x

    futures = [f.remote(i) for i in range(2)]
    print(ray.get(futures))

Put the program into a file named ``debugging.py`` and execute it using:

.. code-block:: bash

    python debugging.py

Each of the 4 executed tasks will drop into a breakpoint when the line

> ``ray.util.pdb.set_trace()`` is executed. You can attach to the debugger by running

Reviewer: Can you say where you can run this command? is it on your laptop, on the head node, on any node?

Author: oh yeah good point, fixed!

.. code-block:: bash

    ray debug

The ``ray debug`` command will print an output like this:

.. code-block:: text

    2020-11-04 15:35:50,011	INFO worker.py:672 -- Connecting to existing Ray cluster at address: 192.168.1.105:6379
    Active breakpoints:
    0: ray::f() | debugging.py:6

    1: ray::f() | debugging.py:6

    Enter breakpoint index or press enter to refresh:

You can now enter ``0`` and hit Enter to jump to the first breakpoint. You will be dropped into PDB
at the break point and can use the ``help`` to see the available actions. Run ``bt`` to see a backtrace

> of the execution:

Reviewer: Can you also have a section here for "Debugger Commands"?

https://docs.python.org/3/library/pdb.html#debugger-commands

Author: I added a section and linked to the to that page. It probably doesn't make sense to replicate it at this point (unless we modify the comments in the future).

.. code-block:: text

    (Pdb) bt
      /Users/pcmoritz/ray/python/ray/workers/default_worker.py(170)<module>()
    -> ray.worker.global_worker.main_loop()
      /Users/pcmoritz/ray/python/ray/worker.py(385)main_loop()
    -> self.core_worker.run_task_loop()
    > /Users/pcmoritz/tmp/debugging.py(7)f()
    -> return x * x

You can inspect the value of ``x`` with ``print(x)``. You can see the current source code with ``ll``
and change stack frames with ``up`` and ``down``. For now let us continue the execution with ``c``.

After the execution is continued, hit ``Control + D`` to get back to the list of break points. Select
the other break point and hit ``c`` again to continue the execution.

The Ray program ``debugging.py`` now finished and should have printed ``[0, 1]``. Congratulations, you
have finished your first Ray debugging session!

Post Mortem Debugging
---------------------

Often we do not know in advance where an error happens, so we cannot set a breakpoint. In these cases,

> we can automatically drop into the debugger when an exception happens.

Reviewer: ```suggestion
we can automatically drop into the debugger when an error occurs or an exception is thrown. This is called *post-mortem debugging*.
```
